### ISSUE TITLE
GATConv costs huge gpu memory

### ISSUE THREAD

**Issue by sw-gong** [rusty1s/pytorch_geometric#527]
Is there any reason for GATConv costing huge GPU memory? My model has only 6 layers of GATConv with around 10k parameters, but it costs more than 32GB GPU memory with 16 batch size. If I replace two of layers with nn.Linear, it would have 4M parameters but only cost 27GB memory. 

It seems there is no broadcasting issues in GATConv implementation. What's the possible issue?

**Comment by rusty1s** [comment]
I will look into it.

**Comment by tformal** [comment]
I also faced similar issues, so I am curious of your thoughts @rusty1s . But in my case, I suspect this is related to my setting where graphs are very dense. This could explain that the memory issue here is not related to the number of parameters, but more likely to the mapping of nodes with x_j/x_i in the message function ?

**Comment by dhorka** [comment]
I think that one source of the problem can be softmax implementation. If I am not wrong, the current pyg implementation of softmax is using autograd to do the backward propagation. As you know, autograd saves a lot of intermidiate values. I think if the deriviative of the softmax is used in the backpropagation will reduce a lot the memory usage.

**Comment by rusty1s** [comment]
Hi, I finally did some tests concerning this issue. IMO, the `softmax` call is not the issue as memory usage stays nearly the same with and without the `softmax` call. This makes sense, as autograd only needs to hold an additional tensor of size `[num_edges, num_heads]` in memory. In addition, overall memory consumption scales linearly with the number of feature channels and the number of edges, so I do not think there is any problem with the implementation.

To be honest, the `GATConv` scales bad by design, as autograd needs to hold tensors of shape `[num_edges, num_heads, 2*num_features]` in memory. There is no real solution to prevent this, as we need to compute attention scores for each edge and each head, based on pair-wise node features. However, I can easily stack up to 10 GAT layers with 8*64 features each on PubMed, and consume about 8GB of memory.

**Comment by sw-gong** [comment]
I see. For my case, it may be caused by large number of edges I guess (around 1.5M directed edges, is this too large?)?

**Comment by rusty1s** [comment]
Yes! Reduce the batch-size ;)

**Comment by sw-gong** [comment]
Do you know what else conv operators won't scale up by edges? It seems that all the conv requires to hold a tensor with the shape at least [num_edges, 2*num_features]?

**Comment by rusty1s** [comment]
All conv operators scale up by edges if they depend on pair-wise node features or edge features. However, something like `GCNConv` can be implemented much more memory-friendly as this operator does not require mapping node features into edge space, *e.g.*, by using sparse-matrix multiplications (with slower runtimes though).
